pallets/offchain-task-scheduler/src/ocw/nonce.rs

```diff
@@ -0,0 +1,244 @@
+			let synced_nonce = StorageValueRef::persistent(key);
+			let synced_nonce = synced_nonce.get::<u64>().unwrap();
+			assert!(synced_nonce.is_none());
+			let nonce = System::<Runtime>::account(acct).nonce;
+			assert_eq!(nonce, 0u64);
+		});
+	}
+
+	#[test]
+	fn parallel_increment() {
+		let (offchain, _) = TestOffchainExt::new();
+		const THREADS: u32 = 3;
+		const LOOP: u32 = 10;
+		let nonces = Arc::new(AtomicU64::new(0));
+
+		let handles = (0..THREADS).into_iter().map(|_| {
+			let offchain = offchain.clone();
+			let nonces = nonces.clone();
+
+			std::thread::spawn(move || {
+				let mut ext_builder = ExtBuilder::default().with_keystore();
+				let acct = {
+					let acct_pubkey = ext_builder.generate_authority();
+					<Runtime as SystemConfig>::AccountId::from(acct_pubkey.into_account().0)
+				};
+				ext_builder.offchain = Some(offchain);
+				ext_builder.with_pool();
+				let mut ext = ext_builder.build();
+
+				let execute = || {
+					roll_to::<Trivial>(1);
+					let call: RuntimeCall =
+						MockTask::Remark(0).forward_task(Runtime::deadline()).expect("call").into();
+
+					for _ in 0..LOOP {
+						assert_ok!(crate::Pallet::<Runtime>::submit_txn_with_synced_nonce(
+							acct.clone(),
+							|_| call.clone(),
+						));
+
+						let nonce = System::<Runtime>::account(acct.clone()).nonce;
+						nonces.fetch_add(nonce, Ordering::Relaxed);
+					}
+				};
+
+				ext.execute_with(execute);
+			})
+		});
+
+		for h in handles {
+			h.join().expect("testing context is shared");
+		}
+
+		let mut ext_builder = ExtBuilder::default();
+		ext_builder.offchain = Some(offchain);
+		let mut ext = ext_builder.build();
+		ext.execute_with(|| {
+			let nonce_post_submition_sum = (THREADS * LOOP) * (THREADS * LOOP + 1) / 2;
+			assert_eq!(nonces.load(Ordering::Relaxed), nonce_post_submition_sum as u64);
+		});
+	}
+
+	#[test]
+	fn lock_works() {
+		let (offchain, _) = TestOffchainExt::new();
+		const THREADS: u32 = 2;
+
+		let handles = (0..THREADS).map(|_| {
+			let offchain = offchain.clone();
+
+			std::thread::spawn(move || {
+				let mut ext_builder = ExtBuilder::default().with_keystore();
+				let acct_pubkey = ext_builder.generate_authority();
+				let acct = <Runtime as SystemConfig>::AccountId::from(acct_pubkey.into_account().0);
+				ext_builder.offchain = Some(offchain);
+				let mut ext = ext_builder.build();
+
+				let execute = || {
+					roll_to::<Trivial>(1);
+
+					let key = lock_key(&acct);
+					let mut lock = crate::Pallet::<Runtime>::nonce_lock_new(&key);
+					let guard = lock.try_lock();
+					guard.map(|g| g.forget()).or_else(|deadline| {
+						// failed to acq guard; move to active guard's deadline boundaries
+						sp_io::offchain::sleep_until(deadline);
+						//deadline still effective
+						lock.try_lock().map(|_| ())
+					})
+				};
+
+				ext.execute_with(execute)
+			})
+		});
+
+		if !handles.into_iter().any(|h| h.join().expect("thread joins").is_err()) {
+			panic!("lock should block")
+		}
```

Comment:
I'm not quite sure what's the intention of this if statement however `panic!`  is not executed here (no coverage).  Can you explain the intent here?

Reply:
You spawn two threads that try to acquire a single lock. One thread will succeed and the other will fail. Assert on at least one thread failing. If the panic is exercised, the locks fails to provide mutual exclusivity.

Reply:
Got it. Thanks!